File: arduinoSlave/pressureplot.ts
import React, { useState, useEffect } from "react";
import Plot from "react-plotly.js";
import io from "socket.io-client"; // Assuming your GCS sends data over socket.io

const PressureGraph = () => {
  const [pressureData, setPressureData] = useState([]);
  const [coefficients, setCoefficients] = useState([]);

  useEffect(() => {
    // Connect to your GCS WebSocket server
    const socket = io("ws://your-gcs-server-address");

    socket.on("pressureData", (data) => {
      // Assuming the data is an array of pressure values from your sensors
      const updatedPressureData = data.map((value) => {
        // Convert pressure data (adjust your formula if needed)
        return 1000 * ((5 / 1023) * value - 2.5);
      });

      setPressureData(updatedPressureData);

      const updatedCoefficients = updatedPressureData.map((pressure) => {
        // Calculate coefficients (adjust the formula based on your requirements)
        return pressure / ((1.225 * 50 ** 2) / 2); // example calculation
      });

      setCoefficients(updatedCoefficients);
    });

    // Clean up the WebSocket connection when the component is unmounted
    return () => {
      socket.disconnect();
    };
  }, []);

  return (
    <div>
      <h1>Pressure Sensor Data</h1>
      <Plot
        data={[
          {
            x: Array.from({ length: pressureData.length }, (_, i) => i + 1), // X-axis: data index
            y: pressureData,
            type: "scatter",
            mode: "lines+markers",
            marker: { color: "red" },
            name: "Pressure Data",
          },
          {
            x: Array.from({ length: coefficients.length }, (_, i) => i + 1),
            y: coefficients,
            type: "scatter",
            mode: "lines+markers",
            marker: { color: "blue" },
            name: "Coefficients",
          },
        ]}
        layout={{
          title: "Pressure Sensor Data and Coefficients",
          xaxis: {
            title: "Sensor Reading Index",
          },
          yaxis: {
            title: "Pressure (Pa)",
          },
        }}
      />
    </div>
  );
};

export default PressureGraph;

File: arduinoSlave/pressureplot.py
import numpy as np
import matplotlib.pyplot as plt
import socketio

# Initialize socket client
sio = socketio.Client()

# Arrays to hold pressure data and calculated coefficients
pressure_data = []
coefficients = []

# Constants
P_max = 400  # Adjust as needed
P_min = 0    # Adjust as needed

# Function to handle incoming data from CubePilot (via WebSocket)
@sio.event
def connect():
    print("Connected to CubePilot")

@sio.event
def disconnect():
    print("Disconnected from CubePilot")

@sio.event
def pressureData(data):
    """
    Process incoming pressure data from CubePilot.
    Data format should be a list of pressure readings.
    """
    global pressure_data, coefficients
    
    # Assuming data is a list of pressure sensor readings
    pressures = np.array(data)

    # Calculate pressure in Pascals using the given formula
    pressure_values = 1000 * ((5 / 1023) * pressures - 2.5)
    pressure_data = np.append(pressure_data, pressure_values)

    # Calculate coefficients using the provided formula
    coeffs = pressure_values / ((1.225 * 50 ** 2) / 2)
    coefficients = np.append(coefficients, coeffs)

    # Plot the pressure data and coefficients
    plot_data()

def plot_data():
    """
    Plot the pressure data and coefficients using matplotlib.
    """
    plt.figure(figsize=(10, 6))
    
    # Plot Pressure Data
    plt.subplot(2, 1, 1)
    plt.plot(pressure_data, color='red', label='Pressure Data')
    plt.title('Pressure Sensor Data')
    plt.xlabel('Sensor Reading Index')
    plt.ylabel('Pressure (Pa)')
    plt.legend()
    
    # Plot Coefficients
    plt.subplot(2, 1, 2)
    plt.plot(coefficients, color='blue', label='Coefficients')
    plt.title('Calculated Coefficients')
    plt.xlabel('Sensor Reading Index')
    plt.ylabel('Coefficient Value')
    plt.legend()
    
    plt.tight_layout()
    plt.pause(0.1)  # This allows the plot to update in real-time

def main():
    # Connect to the WebSocket server (your GCS)
    sio.connect("ws://your-gcs-server-address")

    # Run the plotting in a loop
    plt.ion()  # Interactive mode on
    plt.show()
    try:
        while True:
            pass  # Keep the loop running to receive data and plot in real-time
    except KeyboardInterrupt:
        print("Program terminated.")
        sio.disconnect()

if __name__ == '__main__':
    main()
